network/src/protocols/rpc/mod.rs

```diff
@@ -169,85 +182,197 @@ pub struct Rpc<TSubstream> {
+                        // Handle request with timeout.
+                        let f = async move {
+                            if let Err(err) = tokio::time::timeout(
+                                timeout,
+                                handle_inbound_request(notification_tx, request, peer_handle),
+                            )
+                            .map_err(Into::<RpcError>::into)
+                            .map(|r| r.and_then(|x| x))
+                            .await
+                            {
+                                // Log any errors.
+                                counters::LIBRA_NETWORK_RPC_MESSAGES
+                                    .with_label_values(&["response", "failed"])
+                                    .inc();
+                                warn!(
+                                    "Error handling inbound rpc request from {}: {:?}",
+                                    peer_id_str, err
+                                );
+                            }
+                        };
+                        inbound_rpc_tasks.push(f.boxed());
+                    }
+                    _ => {
+                        error!("Received non-RPC message from Peer actor: {:?}", message);
+                    }
+                }
+            }
+            notif => debug_assert!(
+                false,
+                "Received unexpected event from Peer: {:?}, expected NewMessage",
+                notif
+            ),
+        }
+    }
+
+    /// Handle an outbound rpc request.
+    ///
+    /// Cancellation is done by the client dropping the receiver side of the [`req.res_tx`]
+    /// oneshot channel. If the request is canceled, the rpc future is dropped and the request is
+    /// cancelled. Currently, we don't send a cancellation message to the remote peer.
+    ///
+    /// [`req.res_tx`]: OutboundRpcRequest::res_tx
+    async fn handle_outbound_rpc(
+        &mut self,
+        req: OutboundRpcRequest,
+        outbound_rpc_tasks: &mut FuturesUnordered<BoxFuture<'static, RequestId>>,
+    ) {
+        // If we already have too many pending RPCs, return error immediately.
+        if outbound_rpc_tasks.len() as u32 == self.max_concurrent_outbound_rpcs {
+            warn!(
+                "Pending outbound RPCs ({}) exceeding limit ({}).",
+                outbound_rpc_tasks.len(),
+                self.max_concurrent_outbound_rpcs,
+            );
+            let _result = req.res_tx.send(Err(RpcError::TooManyPending(
+                self.max_concurrent_outbound_rpcs,
+            )));
+            return;
+        }
+
+        // Unpack request.
+        let OutboundRpcRequest {
+            protocol,
+            data: req_data,
+            timeout,
+            mut res_tx,
+            ..
+        } = req;
+
+        let peer_handle = self.peer_handle.clone();
+        let peer_id_str = peer_handle.peer_id().short_str();
```

Comment:
merge into one line? do we need this clone?

Reply:
need to clone because peer_handle needs to moved into a future below on line 294

Reply:
how did I miss that...